File: TS/系统/03．技能系统/00．技能模板+函数/03．技能测试/120．诅咒命中率测试.ts
/** @noSelfInFile */
/**
 * 诅咒命中率测试
 *
 * 输入 "1020"：
 * - 先把玩家1与大法师的“命中率”都清成 0（即默认 100%）
 * - 再对 gg_unit_Hamg_0002 施加 3 秒诅咒
 * - 立即/到期后分别打印单位命中率属性值与“显示命中率”
 *
 * 说明：
 * - YDUserData 的“命中率”是相对 100% 的偏移量
 * - 0   = 100%
 * - 0.2 = 120%
 * - -0.33 = 67%
 */

const jass = require("jass.common") as any;
const g = require("jass.globals") as { gg_unit_Hamg_0002?: any; [key: string]: any };

const { debugLogForce } = require("lib.扩展函数.自定义扩展函数.03．调试输出") as {
  debugLogForce: (this: void, module: string, ...args: any[]) => void;
};
const { 注册聊天命令监听 } = require("系统.00．核心系统.01．事件中心.12．聊天命令事件中心") as {
  注册聊天命令监听: (this: void, 命令: string, 回调: (player: any, command: string) => void) => void;
};

const { YDUserDataGetSafe, YDUserDataSetSafe } = require("lib.扩展函数.YDWE函数.09．YDUserData安全版") as {
  YDUserDataGetSafe: (this: void, tableType: string, tableKey: any, attr: string, valueType: string) => any;
  YDUserDataSetSafe: (this: void, tableType: string, tableKey: any, attr: string, valueType: string, value: any) => void;
};
const { SFB_setCurse } = require("lib.扩展函数.Star扩展函数.Star扩展库.04．快速Buff系统") as {
  SFB_setCurse: (this: void, sourceUnit: any, u: any, time: number) => void;
};
const YDUserDataGet = YDUserDataGetSafe;
const YDUserDataSet = YDUserDataSetSafe;

const GetOwningPlayer = jass.GetOwningPlayer as (whichUnit: any) => any;
const GetPlayerId = jass.GetPlayerId as (whichPlayer: any) => number;
const CreateTimer = jass.CreateTimer as () => any;
const DestroyTimer = jass.DestroyTimer as (whichTimer: any) => void;
const TimerStart = jass.TimerStart as (whichTimer: any, timeout: number, periodic: boolean, handlerFunc: () => void) => void;
const GetExpiredTimer = jass.GetExpiredTimer as () => any;
const GetHandleId = jass.GetHandleId as (whichHandle: any) => number;

const 模块名 = "诅咒命中率测试";
const 测试命令 = "1020";
const ATTR_命中率 = "命中率";
const BUFF_持续时间 = 3.0;
const 到期检查上下文: Record<number, any> = {};

function 归一化实数(this: void, value: any): number {
  if (value == null || value === false || value === "") return 0;
  const n = typeof value === "number" ? value : Number(value);
  return n !== n ? 0 : n;
}

function 读取单位命中率偏移(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return 归一化实数(YDUserDataGet("unit", unit, ATTR_命中率, "real"));
}

function 读取玩家命中率偏移(this: void, player: any): number {
  if (player == null || player === 0) return 0;
  return 归一化实数(YDUserDataGet("player", player, ATTR_命中率, "real"));
}

function 读取有效命中率偏移(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  const unitValue = 读取单位命中率偏移(unit);
  if (unitValue !== 0) return unitValue;
  return 读取玩家命中率偏移(GetOwningPlayer(unit));
}

function 命中率偏移转显示文本(this: void, value: number): string {
  return ((1 + value) * 100).toFixed(0) + "%";
}

function 是否走玩家命中率(this: void, unit: any): boolean {
  const owner = GetOwningPlayer(unit);
  if (owner == null || owner === 0) return false;
  const playerId = GetPlayerId(owner);
  return playerId >= 0 && playerId <= 3;
}

function 打印命中率状态(this: void, stage: string, unit: any): void {
  const owner = GetOwningPlayer(unit);
  const unitOffset = 读取单位命中率偏移(unit);
  const playerOffset = 读取玩家命中率偏移(owner);
  const effectiveOffset = 读取有效命中率偏移(unit);
  const mode = 是否走玩家命中率(unit) ? "玩家" : "单位";
  debugLogForce(
    模块名,
    stage,
    "诅咒目标层=", mode,
    "单位偏移=", unitOffset,
    "玩家偏移=", playerOffset,
    "有效偏移=", effectiveOffset,
    "显示命中率=", 命中率偏移转显示文本(effectiveOffset),
  );
}

function on诅咒到期检查(this: void): void {
  const timer = GetExpiredTimer();
  const timerId = GetHandleId(timer);
  const target = 到期检查上下文[timerId];

  if (target != null && target !== 0) {
    打印命中率状态("诅咒到期后", target);
  }

  delete 到期检查上下文[timerId];
  DestroyTimer(timer);
}

function 安排到期检查(this: void, unit: any, timeout: number): void {
  const timer = CreateTimer();
  const timerId = GetHandleId(timer);
  到期检查上下文[timerId] = unit;
  TimerStart(timer, timeout, false, on诅咒到期检查);
}

function on聊天测试(): void {
  const 大法师 = g.gg_unit_Hamg_0002;
  if (大法师 == null || 大法师 === 0) {
    debugLogForce(模块名, "错误：未找到 gg_unit_Hamg_0002");
    return;
  }

  const owner = GetOwningPlayer(大法师);
  YDUserDataSet("player", owner, ATTR_命中率, "real", 0);
  YDUserDataSet("unit", 大法师, ATTR_命中率, "real", 0);

  打印命中率状态("施加前", 大法师);
  SFB_setCurse(大法师, 大法师, BUFF_持续时间);
  打印命中率状态("诅咒施加后", 大法师);
  安排到期检查(大法师, BUFF_持续时间 + 0.10);

  debugLogForce(模块名, "已施加3秒诅咒，预期：100% -> 67% -> 100%");
}

注册聊天命令监听(测试命令, on聊天测试);
debugLogForce(模块名, "已注册测试：输入", 测试命令, "测试自定义诅咒命中率");

export {};

File: src/系统/03．技能系统/00．技能模板+函数/03．技能测试/120．诅咒命中率测试.lua
local ____lualib = require("lualib_bundle")
local __TS__Number = ____lualib.__TS__Number
local __TS__NumberToFixed = ____lualib.__TS__NumberToFixed
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
--- 诅咒命中率测试
-- 
-- 输入 "1020"：
-- - 先把玩家1与大法师的“命中率”都清成 0（即默认 100%）
-- - 再对 gg_unit_Hamg_0002 施加 3 秒诅咒
-- - 立即/到期后分别打印单位命中率属性值与“显示命中率”
-- 
-- 说明：
-- - YDUserData 的“命中率”是相对 100% 的偏移量
-- - 0   = 100%
-- - 0.2 = 120%
-- - -0.33 = 67%
local jass = require("jass.common")
local g = require("jass.globals")
local ____require_result_0 = require("lib.扩展函数.自定义扩展函数.03．调试输出")
local debugLogForce = ____require_result_0.debugLogForce
local ____require_result_1 = require("系统.00．核心系统.01．事件中心.12．聊天命令事件中心")
local _____6CE8_518C_804A_5929_547D_4EE4_76D1_542C = ____require_result_1["注册聊天命令监听"]
local ____require_result_2 = require("lib.扩展函数.YDWE函数.09．YDUserData安全版")
local YDUserDataGetSafe = ____require_result_2.YDUserDataGetSafe
local YDUserDataSetSafe = ____require_result_2.YDUserDataSetSafe
local ____require_result_3 = require("lib.扩展函数.Star扩展函数.Star扩展库.04．快速Buff系统")
local SFB_setCurse = ____require_result_3.SFB_setCurse
local YDUserDataGet = YDUserDataGetSafe
local YDUserDataSet = YDUserDataSetSafe
local GetOwningPlayer = jass.GetOwningPlayer
local GetPlayerId = jass.GetPlayerId
local CreateTimer = jass.CreateTimer
local DestroyTimer = jass.DestroyTimer
local TimerStart = jass.TimerStart
local GetExpiredTimer = jass.GetExpiredTimer
local GetHandleId = jass.GetHandleId
local _____6A21_5757_540D = "诅咒命中率测试"
local _____6D4B_8BD5_547D_4EE4 = "1020"
local ____ATTR__547D_4E2D_7387 = "命中率"
local ____BUFF__6301_7EED_65F6_95F4 = 3
local _____5230_671F_68C0_67E5_4E0A_4E0B_6587 = {}
local function _____5F52_4E00_5316_5B9E_6570(value)
    if value == nil or value == false or value == "" then
        return 0
    end
    local n = type(value) == "number" and value or __TS__Number(value)
    return n ~= n and 0 or n
end
local function _____8BFB_53D6_5355_4F4D_547D_4E2D_7387_504F_79FB(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return _____5F52_4E00_5316_5B9E_6570(YDUserDataGet("unit", unit, ____ATTR__547D_4E2D_7387, "real"))
end
local function _____8BFB_53D6_73A9_5BB6_547D_4E2D_7387_504F_79FB(player)
    if player == nil or player == 0 then
        return 0
    end
    return _____5F52_4E00_5316_5B9E_6570(YDUserDataGet("player", player, ____ATTR__547D_4E2D_7387, "real"))
end
local function _____8BFB_53D6_6709_6548_547D_4E2D_7387_504F_79FB(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    local unitValue = _____8BFB_53D6_5355_4F4D_547D_4E2D_7387_504F_79FB(unit)
    if unitValue ~= 0 then
        return unitValue
    end
    return _____8BFB_53D6_73A9_5BB6_547D_4E2D_7387_504F_79FB(GetOwningPlayer(unit))
end
local function _____547D_4E2D_7387_504F_79FB_8F6C_663E_793A_6587_672C(value)
    return __TS__NumberToFixed((1 + value) * 100, 0) .. "%"
end
local function _____662F_5426_8D70_73A9_5BB6_547D_4E2D_7387(unit)
    local owner = GetOwningPlayer(unit)
    if owner == nil or owner == 0 then
        return false
    end
    local playerId = GetPlayerId(owner)
    return playerId >= 0 and playerId <= 3
end
local function _____6253_5370_547D_4E2D_7387_72B6_6001(stage, unit)
    local owner = GetOwningPlayer(unit)
    local unitOffset = _____8BFB_53D6_5355_4F4D_547D_4E2D_7387_504F_79FB(unit)
    local playerOffset = _____8BFB_53D6_73A9_5BB6_547D_4E2D_7387_504F_79FB(owner)
    local effectiveOffset = _____8BFB_53D6_6709_6548_547D_4E2D_7387_504F_79FB(unit)
    local mode = _____662F_5426_8D70_73A9_5BB6_547D_4E2D_7387(unit) and "玩家" or "单位"
    debugLogForce(
        _____6A21_5757_540D,
        stage,
        "诅咒目标层=",
        mode,
        "单位偏移=",
        unitOffset,
        "玩家偏移=",
        playerOffset,
        "有效偏移=",
        effectiveOffset,
        "显示命中率=",
        _____547D_4E2D_7387_504F_79FB_8F6C_663E_793A_6587_672C(effectiveOffset)
    )
end
local function ____on_8BC5_5492_5230_671F_68C0_67E5()
    local timer = GetExpiredTimer()
    local timerId = GetHandleId(timer)
    local target = _____5230_671F_68C0_67E5_4E0A_4E0B_6587[timerId]
    if target ~= nil and target ~= 0 then
        _____6253_5370_547D_4E2D_7387_72B6_6001("诅咒到期后", target)
    end
    __TS__Delete(_____5230_671F_68C0_67E5_4E0A_4E0B_6587, timerId)
    DestroyTimer(timer)
end
local function _____5B89_6392_5230_671F_68C0_67E5(unit, timeout)
    local timer = CreateTimer()
    local timerId = GetHandleId(timer)
    _____5230_671F_68C0_67E5_4E0A_4E0B_6587[timerId] = unit
    TimerStart(timer, timeout, false, ____on_8BC5_5492_5230_671F_68C0_67E5)
end
local function ____on_804A_5929_6D4B_8BD5()
    local _____5927_6CD5_5E08 = g.gg_unit_Hamg_0002
    if _____5927_6CD5_5E08 == nil or _____5927_6CD5_5E08 == 0 then
        debugLogForce(_____6A21_5757_540D, "错误：未找到 gg_unit_Hamg_0002")
        return
    end
    local owner = GetOwningPlayer(_____5927_6CD5_5E08)
    YDUserDataSet(
        "player",
        owner,
        ____ATTR__547D_4E2D_7387,
        "real",
        0
    )
    YDUserDataSet(
        "unit",
        _____5927_6CD5_5E08,
        ____ATTR__547D_4E2D_7387,
        "real",
        0
    )
    _____6253_5370_547D_4E2D_7387_72B6_6001("施加前", _____5927_6CD5_5E08)
    SFB_setCurse(_____5927_6CD5_5E08, _____5927_6CD5_5E08, ____BUFF__6301_7EED_65F6_95F4)
    _____6253_5370_547D_4E2D_7387_72B6_6001("诅咒施加后", _____5927_6CD5_5E08)
    _____5B89_6392_5230_671F_68C0_67E5(_____5927_6CD5_5E08, ____BUFF__6301_7EED_65F6_95F4 + 0.1)
    debugLogForce(_____6A21_5757_540D, "已施加3秒诅咒，预期：100% -> 67% -> 100%")
end
_____6CE8_518C_804A_5929_547D_4EE4_76D1_542C(_____6D4B_8BD5_547D_4EE4, ____on_804A_5929_6D4B_8BD5)
debugLogForce(_____6A21_5757_540D, "已注册测试：输入", _____6D4B_8BD5_547D_4EE4, "测试自定义诅咒命中率")
return ____exports
